src/lib/schema.js

```diff
@@ -0,0 +1,147 @@
+import type { ObjectSchema, SchemaRules } from './compose';
+
+import { all, allWhileOK } from './compose';
+
+import {
+  validateIsObject,
+  validateObjHasKey,
+  validateObjPropHasType,
+  validateObjPropPasses,
+  validateObjOnlyHasKeys,
+  validateIsArray,
+  validateArrayItemsHaveType,
+  validateArrayItemsPass,
+  alwaysErr,
+} from './validators';
+
+import { isErr, err, mapErrors, getErrors } from './result';
+
+import { flatten, isObject } from './helpers';
+
+type ConvertToValidator = String => Validator;
+const convertToValidator: ConvertToValidator = key => {
+  if (key === 'required') return validateObjPropHasType('boolean')('required');
+  if (key === 'type') return validateObjPropHasType('string')('type');
+  return validateObjPropHasType('function')('validator');
+};
+
+/**
+ * Validates a set of schema rules as valid
+ */
+type ValidateAsSchemaRules = SchemaRules => Result;
+const validateAsSchemaRules: ValidateAsSchemaRules = rules => {
+  const ruleValidators = Object.keys(rules)
+    .filter(k => ['required', 'type', 'validator'].includes(k))
+    .map(convertToValidator);
+
+  return allWhileOK([validateIsObject, ...ruleValidators])(rules);
+};
+
+/**
+ * Validates a segment of an object schema as a valid set of schema rules
+ */
+type ValidateAsNestedSchemaRules = string => SchemaRules => Result;
+const validateAsNestedSchemaRules: ValidateAsNestedSchemaRules = field => schema =>
+  validateAsSchemaRules(schema[field]);
+
+/**
+ * Enforces a failed validation if the schema is invalid
+ */
+type ProcessSchemaError = SchemaResult => Array<Validator>;
+export const processSchemaError: ProcessSchemaError = schemaResult => [
```

Comment:
Do we only export this for testing? Or do we think it's likely a user of the library would want access to it?